rename package to opencv, add fixes for unix-like platforms, add "virtual packages"

leej3 opened the pull request.

varlackc reviewed (commented):
the license needs to be modified from `BSD 3-clause` to `BSD-3-Clause`
https://github.com/AnacondaRecipes/opencv-feedstock/blob/fix_package_name/recipe/meta.yaml#L251

the `home` and `doc_url` need to use `https` instead of `http`
https://github.com/AnacondaRecipes/opencv-feedstock/blob/fix_package_name/recipe/meta.yaml#L250&&L258

the package is missing the `dev_url`
https://github.com/AnacondaRecipes/opencv-feedstock/blob/fix_package_name/recipe/meta.yaml#L257&&L258

varlackc reviewed (commented):
the package is still failing in `prefect` for `osx-64`

tobijk commented on recipe/meta.yaml:
```diff
@@ -1,5 +1,5 @@
 {% set version = "4.5.4" %}
-{% set use_n_python_patch = os.environ.get('OPENCV_USE_N_PYTHON_PATCH', '0') %}
```
I don't know what this is about but is there maybe a patch file coming with this that needs to be removed or is this just cruft?

leej3 replied:
cruft... it correctly builds with any variant now... I will remove

tobijk commented on recipe/meta.yaml:
```diff
@@ -85,71 +83,168 @@ requirements:
-    {% endfor %}
+
+outputs:
+  - name: opencv
```
It seems odd to me that the libopencv and py-opencv packages would depend on the opencv package and not the other way around. I'm sure there is a good reason, could you add some comments to the recipe explaining the rationale behind these spilts and the dependency handling for other maintainers? Thanks!

leej3 replied:
This is based on a decision to merge libopencv (compiled libraries) and py-opencv (python stack which has dependencies on the compiled libraries). The separation was judged to add excessive maintainence overhead without sufficient benefit. The two additional empty packages are added to keep backwards compatibility with any downstream recipes that made use of either package.

I will confirm that this is indeed implemented as we wish it to be and add an explanatory note for maintainers.

tobijk reviewed (changes requested):
osx-64 failed, I cannot see the logs, is that going to be addressed? Do you have a working Windows build?

leej3 commented:
Thanks for the reviews. I am waiting for access to an intel osx machine and then I will resolve that failing job. I believe the problem arises from the fact that runtime X dependencies are not being supplied my XQuartz in the test environmment. I am not sure how best to resolve such a failure. Once I have figure that out and addressed the points raised above I will remove the WIP label.

skupr-anaconda commented on recipe/meta.yaml:
```diff
@@ -85,71 +87,172 @@ requirements:
+      skip: True  # [(win and py27) or s390x]
+    requirements:
+      host:
+        - {{ pin_subpackage('opencv', max_pin='x') }}
```
For the python output `py-opencv` please add explicitly in the `host`:
- `setuptools`, see https://github.com/opencv/opencv/blob/4223495e6cd67011f86b8ecd9be1fa105018f3b1/modules/python/package/setup.py#L4
- `wheel`

skupr-anaconda commented on recipe/meta.yaml:
```diff
@@ -85,71 +87,172 @@ requirements:
+        - {{ pin_subpackage('opencv', max_pin='x') }}
 
 about:
   home: http://opencv.org/
```
The license is `Apache-2.0` for **OpenCV 4.5.0+**, https://github.com/opencv/opencv/blob/4.5.4/LICENSE and https://opencv.org/license/

skupr-anaconda replied:
And `license_family` should be `Apache`

leej3 replied:
Yep, still working on it... not ready for review yet. I marked it as work in progress (WIP). Is there a convention you guys use for that? I could convert it to a draft on the github interface perhaps.

leej3 commented:
+ The failure on OSX was due to a set of missing shared libraries. These are not built because libprotobuf 3.5.1 is not currently detected in the OSX. 
+ I have updated all of the metadata as requested (excepting setuptools/wheel for py-opencv which is not appropriate).
+ Python 3.10 on aarch64 is not included because qt 5.15.2 is missing with this variant on that architecture. I will update this recipe once that is remedied (I will be working on Qt next so perhaps by me).
+ The changes required for the Windows build were extensive and I will submit them for separate review over the coming day.

katietz reviewed (approved):
thanks

leej3 commented:
Sorry, updated with what I thought would be a harmless change... turns out it is not so harmless. I have reverted to the previous commit.
